src/subsplit_dag_node.hpp

```diff
@@ -18,112 +18,231 @@
-  enum class Direction : bool {
-    Rootward = 0,
-    Leafward = 1,
-  };
-  enum class ParentClade : bool {
-    Left = 0,
-    Right = 1,
-  };
 
-  SubsplitDAGNode(size_t id, Bitset subsplit)
-      : id_(id), subsplit_(std::move(subsplit)) {}
+  GenericSubsplitDAGNode(T& node) : node_{node} {}
+  GenericSubsplitDAGNode(const GenericSubsplitDAGNode<std::remove_const_t<T>>& other) : node_{other.node_} {}
 
-  size_t Id() const { return id_; }
-  const Bitset &GetBitset() const { return subsplit_; }
+  size_t Id() const { return node_.GetId(); }
+  const Bitset &GetBitset() const { return node_.GetSubsplit(); }
   const Bitset GetBitset(bool rotated) const {
-    return rotated ? subsplit_.SubsplitRotate() : subsplit_;
+    return rotated ? node_.GetSubsplit().SubsplitRotate() : node_.GetSubsplit();
   }
 
   // ** Node Types
 
   // Is the node the DAG root (universal ancestor of the DAG)?
   bool IsDAGRootNode() const {
-    return (right_rootward_.empty() && left_rootward_.empty());
+    return (GetRightRootward().empty() && GetLeftRootward().empty());
   }
   // Is the node a rootsplit (direct descendent of root, dividing entire taxon set)?
-  bool IsRootsplit() const { return subsplit_.SubsplitIsRootsplit(); }
+  bool IsRootsplit() const { return node_.GetSubsplit().SubsplitIsRootsplit(); }
   // Is the node a leaf (has no descendents)?
-  bool IsLeaf() const { return left_leafward_.empty() && right_leafward_.empty(); }
+  bool IsLeaf() const { return GetLeftLeafward().empty() && GetRightLeafward().empty(); }
 
   // ** Edges
 
   // Add edge from this node to adjacent_node.
-  void AddEdge(size_t adjacent_node_id, bool is_leafward, bool is_rotated) {
-    if (is_leafward) {
-      is_rotated ? AddLeftLeafward(adjacent_node_id)
-                 : AddRightLeafward(adjacent_node_id);
-    } else {
-      is_rotated ? AddLeftRootward(adjacent_node_id)
-                 : AddRightRootward(adjacent_node_id);
-    }
+  void AddEdge(size_t adjacent_node_id, LineId edge_id, Direction which_direction,
+               Clade which_clade) {
+    node_.AddNeighbor(which_direction, which_clade, adjacent_node_id, edge_id);
   }
-  void AddEdge(size_t adjacent_node_id, Direction which_direction,
-               ParentClade which_clade) {
-    bool is_leafward = (which_direction == Direction::Leafward);
-    bool is_rotated = (which_clade == ParentClade::Left);
-    AddEdge(adjacent_node_id, is_leafward, is_rotated);
-  }
-  void AddLeftLeafward(size_t node_id) { left_leafward_.push_back(node_id); }
-  void AddRightLeafward(size_t node_id) { right_leafward_.push_back(node_id); }
-  void AddLeftRootward(size_t node_id) { left_rootward_.push_back(node_id); }
-  void AddRightRootward(size_t node_id) { right_rootward_.push_back(node_id); }
 
   // Get vector of all adjacent node vectors along the specified direction.
-  const SizeVector &GetEdge(bool is_leafward, bool is_rotated) {
+  ConstNeighborsView GetEdge(bool is_leafward, bool is_rotated) const {
```

Comment:
Interesting that this returns a NeighborsView. I see how this is a nice consistent choice, but does it make it difficult to consume?

Reply:
[Const]NeighborsView looks to the users similarly to a SizeVector - can be iterated, and queried for size() and empty(). It can easily be extended to support indexed access with [i] and at(i) like LinesView. It wasn't needed so far by the rest of the code so I didn't add those, but I'd like to add them in the future.

Reply:
This looks like just what is needed for the documentation suggested above.